keps/sig-auth/2843-request-ctx-in-token-review/README.md

```diff
@@ -0,0 +1,109 @@
+# KEP-2843: Forward Additional Request Metadata in TokenReviews
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes enhancing the TokenReview API to convey additional request
+context to the backing Identity Provider (IDP).
+
+## Motivation
+
+Today, certain request attributes that are known to kube-apiserver may be
+relevant to the backend of a TokenReview webhook. For example:
+
+* Was the request made over https?
+* What are the IP addresses of the peer and origin of the request?
+
+These attributes can be used to enhance audit logs in the IDP, determine access
+levels (such as whether the request originates from a trusted network), or
+enforce security policy centrally (such as disallowing bearer token
+authentication over insecure protocols).
+
+### Goals
+
+* Forward additional request metadata to TokenReview. 
+
+### Non-Goals
+
+* Forward resource metadata to TokenReview: The object of the request (e.g.
+  which verb, which namespace, which pod) should remain unknown to the IDP.
+* Forward addition request metadata to SubjectAccessReview or AdmissionReview:
+  That's what User.ExtraInfo is for.
```

Comment:
> That's what User.ExtraInfo is for.

is it? do attributes of the incoming request belong to the user? a side effect of including here is that it gets persisted in places like CSR `spec.extra`... is that desired?

Reply:
Trying to keep this minimal, I imagined a few scenarios that are satisfied by only plumbing this to TokenReview:
* Auditing IPs in the IDP
* 401 rejecting based on some policy with an input of attributes + token.
* Forwarding (hopefully summarized, not blindly copied) metadata to authorization and admission via extra info.

I buy the argument for extending the plumbing of the raw metadata through authorization to support users authenticated via different mechanisms. Is there a reasonable case to expose this info to admission? What's the philosophy there?

Reply:
>  Is there a reasonable case to expose this info to admission? What's the philosophy there?

* in general, I would expect admission webhooks to have all the information given to authorization webhooks
* implementing targeted deny-only authorization checks seems better done with a dynamically-configured resource-scoped admission webhook than with a statically-configured authorization webhook
* any policy that needs to make decisions based on request content and request origin would have to do so in admission

Reply:
Updated, PTAL.